php-frontend/src/main/java/org/sonar/php/tree/impl/declaration/EnumDeclarationTreeImpl.java

```diff
@@ -21,71 +21,44 @@
 import org.sonar.php.utils.collections.IteratorUtils;
 import org.sonar.plugins.php.api.tree.Tree;
+import org.sonar.plugins.php.api.tree.declaration.AttributeGroupTree;
+import org.sonar.plugins.php.api.tree.declaration.ClassMemberTree;
 import org.sonar.plugins.php.api.tree.declaration.EnumDeclarationTree;
+import org.sonar.plugins.php.api.tree.declaration.NamespaceNameTree;
 import org.sonar.plugins.php.api.tree.expression.NameIdentifierTree;
 import org.sonar.plugins.php.api.tree.lexical.SyntaxToken;
 import org.sonar.plugins.php.api.tree.statement.EnumCaseTree;
-import org.sonar.plugins.php.api.visitors.VisitorCheck;
 
-public class EnumDeclarationTreeImpl extends PHPTree implements EnumDeclarationTree {
+public class EnumDeclarationTreeImpl extends ClassDeclarationTreeImpl implements EnumDeclarationTree {
 
-  private final SyntaxToken enumToken;
-  private final NameIdentifierTree name;
-  private final SyntaxToken openCurlyBraceToken;
   private final List<EnumCaseTree> cases;
-  private final SyntaxToken closeCurlyBraceToken;
 
-  public EnumDeclarationTreeImpl(SyntaxToken enumToken, NameIdentifierTree name, SyntaxToken openCurlyBraceToken,
-    List<EnumCaseTree> cases, SyntaxToken closeCurlyBraceToken) {
-    this.enumToken = enumToken;
-    this.name = name;
-    this.openCurlyBraceToken = openCurlyBraceToken;
-    this.cases = cases;
-    this.closeCurlyBraceToken = closeCurlyBraceToken;
-  }
-
-  @Override
-  public SyntaxToken enumToken() {
-    return enumToken;
-  }
-
-  @Override
-  public NameIdentifierTree name() {
-    return name;
-  }
-
-  @Override
-  public SyntaxToken openCurlyBraceToken() {
-    return openCurlyBraceToken;
+  public EnumDeclarationTreeImpl(List<AttributeGroupTree> attributeGroups, SyntaxToken enumToken, NameIdentifierTree name,
+    @Nullable InternalSyntaxToken implementsToken, SeparatedListImpl<NamespaceNameTree> superInterfaces, SyntaxToken openCurlyBraceToken,
+    List<ClassMemberTree> members, SyntaxToken closeCurlyBraceToken) {
+    super(Kind.ENUM_DECLARATION, attributeGroups, null, enumToken, name, null, null,
+      implementsToken, superInterfaces, openCurlyBraceToken, members, closeCurlyBraceToken);
+    this.cases = members.stream().filter(m -> m.is(Kind.ENUM_CASE)).map(EnumCaseTree.class::cast).collect(Collectors.toList());
   }
 
   @Override
   public List<EnumCaseTree> cases() {
     return cases;
   }
 
-  @Override
-  public SyntaxToken closeCurlyBraceToken() {
-    return closeCurlyBraceToken;
-  }
-
   @Override
   public Iterator<Tree> childrenIterator() {
-    return IteratorUtils.concat(IteratorUtils.iteratorOf(enumToken, name, openCurlyBraceToken),
-      cases.iterator(),
-      IteratorUtils.iteratorOf(closeCurlyBraceToken));
-  }
-
-  @Override
-  public void accept(VisitorCheck visitor) {
-    visitor.visitEnumDeclaration(this);
-  }
-
-  @Override
-  public Kind getKind() {
-    return Kind.ENUM_DECLARATION;
+    return IteratorUtils.concat(attributeGroups().iterator(),
```

Comment:
Technically, in this PR, `childrenIterator` does not have to be overridden yet. Using the one from the parent would work too. The only advantage now is that we avoid the `null` values in the iterator for stuff like super classes which cannot appear for enums.
Also it will need to be overridden anyway in the next ticket SONARPHP-1258 to add the children representing the backed type in something like `enum Foo: string {}`.